File: backend/app/schemas/service_request.py
from __future__ import annotations
from typing import Optional, List
from pydantic import BaseModel, Field
from datetime import datetime
from .user import UserRead
from .pet import PetRead

class ServiceRequestBase(BaseModel):
    service_type: str = Field(..., description="Type of service requested")
    title: str = Field(..., min_length=5, max_length=100, description="Request title")
    description: str = Field(..., min_length=10, max_length=1000, description="Detailed description")
    pet_ids: List[int] = Field(..., min_items=1, description="Pet IDs to share with providers")
    location: Optional[str] = Field(None, max_length=200, description="Preferred location")
    preferred_dates: Optional[List[str]] = Field(None, description="Preferred dates")
    budget_min: Optional[int] = Field(None, ge=0, description="Minimum budget")
    budget_max: Optional[int] = Field(None, ge=0, description="Maximum budget")
    experience_years_min: Optional[int] = Field(None, ge=0, le=50, description="Minimum experience required")
    languages: Optional[List[str]] = Field(None, description="Required languages")
    special_requirements: Optional[str] = Field(None, max_length=500, description="Special requirements")
    is_urgent: bool = Field(False, description="Is this request urgent?")

class ServiceRequestCreate(ServiceRequestBase):
    pass

class ServiceRequestUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=5, max_length=100)
    description: Optional[str] = Field(None, min_length=10, max_length=1000)
    location: Optional[str] = Field(None, max_length=200)
    preferred_dates: Optional[List[str]] = None
    budget_min: Optional[int] = Field(None, ge=0)
    budget_max: Optional[int] = Field(None, ge=0)
    experience_years_min: Optional[int] = Field(None, ge=0, le=50)
    languages: Optional[List[str]] = None
    special_requirements: Optional[str] = Field(None, max_length=500)
    is_urgent: Optional[bool] = None
    status: Optional[str] = Field(None, pattern="^(open|in_progress|completed|closed)$")

class ServiceRequestRead(ServiceRequestBase):
    id: int
    user_id: int
    assigned_provider_id: Optional[int] = None
    status: str
    views_count: int
    responses_count: int
    created_at: datetime
    updated_at: datetime
    expires_at: Optional[datetime]
    
    # Relationships
    user: Optional[UserRead] = None
    assigned_provider: Optional[UserRead] = None
    pets: Optional[List[PetRead]] = None
    
    class Config:
        from_attributes = True

class ServiceRequestSummary(BaseModel):
    """Summary view for browsing requests"""
    id: int
    title: str
    service_type: str
    location: Optional[str]
    budget_min: Optional[int]
    budget_max: Optional[int]
    is_urgent: bool
    created_at: datetime
    views_count: int
    responses_count: int
    user: UserRead
    pets: List[PetRead]
    
    class Config:
        from_attributes = True

File: frontend/src/types/services/serviceRequest.ts
import type { Pet } from '../pets';
import type { User } from '../auth';

export type ServiceRequestStatus = 'open' | 'in_progress' | 'completed' | 'closed';

export interface ServiceRequest {
  id: number;
  user_id: number;
  assigned_provider_id?: number;
  service_type: string;
  title: string;
  description: string;
  pet_ids: number[];
  location?: string;
  preferred_dates?: string[];
  budget_min?: number;
  budget_max?: number;
  experience_years_min?: number;
  languages?: string[];
  special_requirements?: string;
  status: ServiceRequestStatus;
  is_urgent: boolean;
  views_count: number;
  responses_count: number;
  created_at: string;
  updated_at: string;
  expires_at?: string;
  
  // Relationships
  user?: User;
  assigned_provider?: User;
  pets?: Pet[];
}

export interface ServiceRequestCreate {
  service_type: string;
  title: string;
  description: string;
  pet_ids: number[];
  location?: string;
  preferred_dates?: string[];
  budget_min?: number;
  budget_max?: number;
  experience_years_min?: number;
  languages?: string[];
  special_requirements?: string;
  is_urgent?: boolean;
}

export interface ServiceRequestUpdate {
  title?: string;
  description?: string;
  location?: string;
  preferred_dates?: string[];
  budget_min?: number;
  budget_max?: number;
  experience_years_min?: number;
  languages?: string[];
  special_requirements?: string;
  is_urgent?: boolean;
  status?: ServiceRequestStatus;
}

export interface ServiceRequestSummary {
  id: number;
  title: string;
  service_type: string;
  location?: string;
  budget_min?: number;
  budget_max?: number;
  is_urgent: boolean;
  created_at: string;
  views_count: number;
  responses_count: number;
  user: User;
  pets: Pet[];
}

export interface ServiceRequestFilters {
  service_type?: string;
  location?: string;
  budget_min?: number;
  budget_max?: number;
  is_urgent?: boolean;
  limit?: number;
  offset?: number;
}